### PR TITLE
algo client side rate limiting

### PR DESCRIPTION
**Pull request by DanBurton**

### CODE REVIEW

File: js/stdlib/ts/ALGO.ts [hunk 1]
```diff
@@ -266,6 +269,46 @@ type AlgodTxn = {
...
+// module-wide config
+export var customHttpEventHandler: (e: RHC.Event) => Promise<unknown> = async () => null
+/**
+ * @description client-side rate limiting.
+ *  Setting this to any positive number will also prevent requests from being sent in parallel.
+ *  Rate limiting is applied to all outgoing http requests, even if they are to different servers.
+ */
+export var minMillisBetweenRequests: number = 0;
+const reqLock = new Lock();
+var currentReqNum: number | undefined = undefined;
```
Comment (jeapostrophe) [inline js/stdlib/ts/ALGO.ts:281]: `let` not `var`. Var is weird

File: js/stdlib/ts/ALGO_ReachHTTPClient.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    relativePath: string,
+    data: Uint8Array,
+    query?: Query<string>,
+    requestHeaders?: Record<string, string>
+    ): Promise<BaseHTTPClientResponse> {
+    const method = 'delete';
+    const eventDetails: EventDetails = { method, relativePath, data, query, requestHeaders }
+    return await this._doTheThing(eventDetails, async () => await this.bc.delete(relativePath, data, query, requestHeaders));
+  }
+}
```
Comment (jeapostrophe) [inline js/stdlib/ts/ALGO_ReachHTTPClient.ts:80]: This code is asking for abstraction of `this.bc["delete"]` etc

File: js/stdlib/ts/ALGO_UTBC.ts [hunk 3]
```diff
@@ -0,0 +1,179 @@
+// @ts-nocheck
+// XXX get them to export this. Webpack can't pluck the source file so we copy/paste it instead.
```
Comment (jeapostrophe) [inline js/stdlib/ts/ALGO_UTBC.ts:2]: Please make an issue or PR to `js-algorand-sdk`

File: js/stdlib/ts/ETH_like.ts [hunk 4]
```diff
@@ -1111,6 +1111,14 @@ function unsafeGetMnemonic(acc: Account|NetworkAccount): string {
   }
 }
 
+function setMinMillisBetweenRequests() {
+  console.warn(`setMinMillisBetweenRequests is not yet implemented on this connector`);
```
Comment (jeapostrophe) [inline js/stdlib/ts/ETH_like.ts:1115]: I think it should "not supported"

File: js/stdlib/ts/ALGO_ReachHTTPClient.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  async get(
+    relativePath: string,
+    query?: Query<string>,
+    requestHeaders?: Record<string, string>
+    ): Promise<BaseHTTPClientResponse> {
+    const method = 'get';
+    const eventDetails: EventDetails = { method, relativePath, query, requestHeaders};
+    return await this._doTheThing(eventDetails, async () => await this.bc.get(relativePath, query, requestHeaders));
```
Comment (jeapostrophe) [inline js/stdlib/ts/ALGO_ReachHTTPClient.ts:56]: Even this could be abstracted with `...(data ? [data] : [])` and then you got one function for 'em all

**Review by jeapostrophe — CHANGES_REQUESTED** [review]

**Review by jeapostrophe — CHANGES_REQUESTED** [review]
I realized that adding it to ETH_like doesn't actually add it to ETH and CFX, because there's more to do there. You should add these functions to interfaces.ts and see the error.

Also, docs

**Comment by DanBurton** [comment]
The changes to interfaces.ts, ETH.ts, and CFX.ts are already in the PR. Working on docs now.

**Comment by jeapostrophe** [comment]
Ah, I was looking at the changes, not the full thing. Sorry!

**Review by jeapostrophe — APPROVED** [review]

**Pull request merged**